File: generated_code/GeminiPro/Java/034_SplayTree.java
/**
 * A complete, self-contained Java class that implements a Splay Tree for an integer set.
 * A Splay Tree is a self-balancing binary search tree with the property that recently
 * accessed elements are quick to access again. It performs basic operations such as
 * insertion, deletion, and search in O(log n) amortized time.
 */
public class SplayTree {

    /**
     * Represents a node in the Splay Tree.
     * It is a static inner class as it does not need to access instance members of SplayTree.
     */
    private static class Node {
        int key;
        Node left;
        Node right;

        /**
         * Constructor for a Node.
         * @param key The integer key for the node.
         */
        Node(int key) {
            this.key = key;
            this.left = null;
            this.right = null;
        }
    }

    private Node root;

    /**
     * Constructs an empty Splay Tree.
     */
    public SplayTree() {
        this.root = null;
    }

    // --- Core Splaying Logic and Rotations ---

    /**
     * Performs a right rotation on the subtree rooted at node x.
     * @param x The root of the subtree to rotate.
     * @return The new root of the subtree after rotation.
     */
    private Node rightRotate(Node x) {
        Node y = x.left;
        x.left = y.right;
        y.right = x;
        return y;
    }

    /**
     * Performs a left rotation on the subtree rooted at node x.
     * @param x The root of the subtree to rotate.
     * @return The new root of the subtree after rotation.
     */
    private Node leftRotate(Node x) {
        Node y = x.right;
        x.right = y.left;
        y.left = x;
        return y;
    }

    /**
     * Splays the tree for a given key. This operation moves the node with the given key
     * (or the last accessed node on the search path if the key is not found) to the root.
     * @param node The root of the current subtree.
     * @param key The key to splay for.
     * @return The new root of the tree after splaying.
     */
    private Node splay(Node node, int key) {
        // Base case: node is null or key is present at root
        if (node == null || node.key == key) {
            return node;
        }

        if (key < node.key) {
            // Key is in the left subtree
            if (node.left == null) {
                return node; // Key not found, return last accessed node
            }
            // Zig-Zig (Left-Left)
            if (key < node.left.key) {
                node.left.left = splay(node.left.left, key);
                node = rightRotate(node);
            } 
            // Zig-Zag (Left-Right)
            else if (key > node.left.key) {
                node.left.right = splay(node.left.right, key);
                if (node.left.right != null) {
                    node.left = leftRotate(node.left);
                }
            }
            // Do second rotation for Zig-Zig/Zig-Zag, or the only rotation for Zig
            return (node.left == null) ? node : rightRotate(node);
        } else { // key > node.key
            // Key is in the right subtree
            if (node.right == null) {
                return node; // Key not found, return last accessed node
            }
            // Zig-Zig (Right-Right)
            if (key > node.right.key) {
                node.right.right = splay(node.right.right, key);
                node = leftRotate(node);
            } 
            // Zig-Zag (Right-Left)
            else if (key < node.right.key) {
                node.right.left = splay(node.right.left, key);
                if (node.right.left != null) {
                    node.right = rightRotate(node.right);
                }
            }
            // Do second rotation for Zig-Zig/Zig-Zag, or the only rotation for Zig
            return (node.right == null) ? node : leftRotate(node);
        }
    }

    // --- Public API Methods ---

    /**
     * Searches for a key in the tree. If the key is found, it is splayed
     * to the root. If not found, the last accessed node is splayed to the root.
     * @param key The key to search for.
     * @return The key as an Integer if found, otherwise null.
     */
    public Integer search(int key) {
        if (root == null) {
            return null;
        }
        root = splay(root, key);
        if (root.key == key) {
            return key;
        }
        return null;
    }

    /**
     * Inserts a key into the Splay Tree. As this is a set, duplicates are not allowed.
     * After insertion, the new node becomes the root of the tree.
     * @param key The key to insert.
     */
    public void insert(int key) {
        // If the tree is empty, create a new root
        if (root == null) {
            root = new Node(key);
            return;
        }

        // Splay the tree for the key. The node closest to the key will become the root.
        root = splay(root, key);

        // If key is already present, do nothing.
        if (root.key == key) {
            return;
        }

        Node newNode = new Node(key);

        // The new node becomes the new root. The old root and one of its subtrees
        // become children of the new node.
        if (key < root.key) {
            newNode.right = root;
            newNode.left = root.left;
            root.left = null;
        } else { // key > root.key
            newNode.left = root;
            newNode.right = root.right;
            root.right = null;
        }
        root = newNode;
    }

    /**
     * Deletes a key from the Splay Tree.
     * @param key The key to delete.
     */
    public void delete(int key) {
        if (root == null) {
            return;
        }

        // Splay the tree for the key. If the key exists, it becomes the root.
        root = splay(root, key);

        // If key is not in the tree, do nothing.
        if (root.key != key) {
            return;
        }

        // Key is at the root. Now perform the deletion.
        Node leftSubtree = root.left;
        Node rightSubtree = root.right;

        if (leftSubtree == null) {
            // If no left child, the right subtree becomes the new tree.
            root = rightSubtree;
        } else {
            // Make the left subtree the new root.
            root = leftSubtree;
            // Splay the new tree for the key we just deleted. This brings the
            // largest element (predecessor) of the left subtree to its root.
            // This new root is guaranteed to have a null right child.
            root = splay(root, key);
            // Attach the original right subtree to the new root.
            root.right = rightSubtree;
        }
    }
}

File: generated_code/GeminiPro/Python/034_SplayTree.py
import sys

# It's recommended to increase recursion limit for deep trees, though splaying mitigates this.
# sys.setrecursionlimit(2000)

class SplayTree:
    """
    Implements a splay tree for storing unique integers.

    This data structure behaves like a set and supports insert, delete, and search operations.
    The key feature is that any accessed (searched, inserted, or deleted) node
    is moved to the root of the tree through a series of rotations,
    a process called "splaying". This keeps frequently accessed elements
    near the top for faster subsequent access.
    """

    class _Node:
        """A node in the splay tree."""
        def __init__(self, key):
            self.key = key
            self.parent = None
            self.left = None
            self.right = None

    def __init__(self):
        """Initializes an empty splay tree."""
        self.root = None

    def _left_rotate(self, x):
        """Performs a left rotation on node x."""
        y = x.right
        x.right = y.left
        if y.left:
            y.left.parent = x

        y.parent = x.parent
        if not x.parent:
            self.root = y
        elif x == x.parent.left:
            x.parent.left = y
        else:
            x.parent.right = y

        y.left = x
        x.parent = y

    def _right_rotate(self, x):
        """Performs a right rotation on node x."""
        y = x.left
        x.left = y.right
        if y.right:
            y.right.parent = x

        y.parent = x.parent
        if not x.parent:
            self.root = y
        elif x == x.parent.right:
            x.parent.right = y
        else:
            x.parent.left = y

        y.right = x
        x.parent = y

    def _splay(self, node):
        """
        Splays the given node to the root of the tree.
        """
        while node.parent:
            p = node.parent
            gp = p.parent
            if not gp:  # Zig case
                if node == p.left:
                    self._right_rotate(p)
                else:
                    self._left_rotate(p)
            elif node == p.left and p == gp.left:  # Zig-Zig case (left-left)
                self._right_rotate(gp)
                self._right_rotate(p)
            elif node == p.right and p == gp.right:  # Zig-Zig case (right-right)
                self._left_rotate(gp)
                self._left_rotate(p)
            elif node == p.right and p == gp.left:  # Zig-Zag case (left-right)
                self._left_rotate(p)
                self._right_rotate(gp)
            else:  # Zig-Zag case (right-left)
                self._right_rotate(p)
                self._left_rotate(gp)

    def search(self, key):
        """
        Searches for a key in the splay tree.

        Performs the splaying operation on the accessed node if found,
        or on its would-be parent if not found.

        Args:
            key: The integer key to search for.

        Returns:
            True if the key is found, False otherwise.
        """
        node = self.root
        last_visited = None
        while node:
            last_visited = node
            if key < node.key:
                node = node.left
            elif key > node.key:
                node = node.right
            else:
                # Key found, splay the node and return
                self._splay(node)
                return True
        
        # Key not found. Splay the last visited node (the parent) if it exists.
        if last_visited:
            self._splay(last_visited)
        return False

    def insert(self, key):
        """
        Inserts a key into the splay tree.

        If the key already exists, the existing node is splayed to the root.
        If the key is new, it is inserted and then the new node is splayed to the root.

        Args:
            key: The integer key to insert.
        """
        # Find position for new node, similar to BST insert
        parent = None
        current = self.root
        while current:
            parent = current
            if key < current.key:
                current = current.left
            elif key > current.key:
                current = current.right
            else:
                # Key already exists, splay it and we are done
                self._splay(current)
                return

        # Create the new node
        new_node = self._Node(key)
        new_node.parent = parent

        if not parent:
            # The tree was empty
            self.root = new_node
        elif new_node.key < parent.key:
            parent.left = new_node
        else:
            parent.right = new_node

        # Splay the newly inserted node to the root
        self._splay(new_node)

    def delete(self, key):
        """
        Deletes a key from the splay tree.

        The node (if found) is first splayed to the root, then removed.
        If the key is not found, the last accessed node in the search path is splayed.

        Args:
            key: The integer key to delete.
        """
        # Splay the node with the key (or its parent) to the root
        if not self.search(key):
            # Key not in the tree. search() already splayed the closest node.
            return

        # At this point, the node to delete is the root because search(key) found it
        # and splayed it.
        z = self.root
        
        left_subtree = z.left
        right_subtree = z.right

        if not left_subtree:
            # No left child, so the right subtree becomes the new tree.
            self.root = right_subtree
            if self.root:
                self.root.parent = None
        elif not right_subtree:
            # No right child, so the left subtree becomes the new tree.
            self.root = left_subtree
            if self.root:
                self.root.parent = None
        else:
            # Both children exist. We need to join the left and right subtrees.
            # 1. Make the left subtree the main tree temporarily.
            left_subtree.parent = None
            self.root = left_subtree

            # 2. Find the maximum element in this new tree (the new root).
            max_node = self.root
            while max_node.right:
                max_node = max_node.right

            # 3. Splay this maximum element to the root.
            #    After this, self.root will be max_node, and it has no right child.
            self._splay(max_node)

            # 4. Attach the original right subtree as the right child of the new root.
            self.root.right = right_subtree
            right_subtree.parent = self.root